RFCs/0000-pdk-template-v2.md

```diff
@@ -0,0 +1,566 @@
+
+Using the example Litmus, Gemfile, Appveyor and Travis templates.
+
+``` text
+                   Shared Setting
+    Publish     +------------------+                       Subscribe
+       +------> | use_litmus: true | +--------+---------------+---------------+
+       |        +------------------+          |               |               |
+       |                                      v               v               v
++------+-----+                         +------------+  +-------------+  +------------+
+|   litmus   |                         |  gemfile   |  | appveyor-ci |  |  travis-ci |
+|  template  |                         |  template  |  |  template   |  |  template  |
++------------+                         +------------+  +-------------+  +------------+
+```
+
+1) The `litmus` template publishes a shared setting called `use_litmus` (with a value of true)
+
+2) The `gemfile` template subscribes to this setting, and modifies it's internal settings to add the litmus gem to the `Gemfile` file.
+
+3) The `appveyor-ci` template subscribes to this setting, and modifies it's internal settings to add the litmus tasks to the `appveyor.yml` file.
+
+4) The `travis-ci` template subscribes to this setting, and modifies it's internal settings to add the litmus tasks to the `.travis.yml` file.
+
+Note - The name of the setting is used to convey intent, not a particular implementation. In the example above, it conveys the intent that other templates should use litmus. Instead of, say, `add_litmus_gem` which only makes sense for the gemfile template, but not appveyor.
+
+Note - The template will define the type of object the setting can contain. In the example above the object type would be `boolean`. Another example could be: A `puppet-lint` template could publish an `available_puppet-lint_plugins` shared setting with a value of `["puppet-lint-trailing_newline-check", "puppet-lint-roles_and_profiles-check"]`.  This would have an object type of `array[string]`.
+
+Note - Templates would be able to subscribe to more than one setting.  For example the `gemfile` template would subscribe to the `use_litmus` and `available_puppet-lint_plugins` settings and add the required gems to the templated `Gemfile`.
+
+### Template Engine Lifecycle and Template hook files
+
+Most simple templates will only require a setting and an ERB file for rendering however there are cases where more advanced logic is required to be used.  Either because putting that logic _into_ an ERB file results in a horrible mess which is hard to maintain, or it's just not possible to express the logic in ERB. The V2 Template Engine will allow Template Authors to execute Ruby files at various parts of the templating life cycle using hooks.  These files will be located in the template directory and have specific names, reflecting their state in the lifecycle.
+
+#### Template Engine Lifecycle
```

Comment:
Note - We should note which hooks would be the most common/most needed and document them fully